### PR TITLE
Conditionally exclude some L0 input files in size amp compaction

### PR DESCRIPTION
**Pull request by hx235**
**Context/Summary:**
A size amp compaction can select and prevent a large number of L0 files from being selected by other compaction. If such compaction is running long or being queued behind, these L0 files will exist for long. With a few more flushes, we can run into write stop triggered by # L0 files. We've seen this happen on a host with many DBs sharing same thread pool, each of these DBs submits a size amp compaction with (110-180)+ files to the pool upon reopen and with a few more flushes, they hit the 200 L0 write stop condition.

The idea is to exclude some L0 input files in size amp compaction that are harmless to size amp reduction but improve the situation described above.

The exclusion algorithm is in `MightExcludeNewL0sToReduceWriteStop()` with two elements:

1.  #L0 to exclude + (level0_stop_writes_trigger - num_l0_input_pre_exclusion) should be in the range of [min_merge_width, max_merge_width]. 
    -  This is to ensure we are excluding enough L0 input files but not too many to be qualified to picked for another compaction along with the incoming future L0 files before write stop.
2. Based on (1), further constrain #L0 to exclude based on the post-exclusion compaction score. The goal is to ensure our exclusion will not disqualify the size amp compaction from being a size amp compaction after exclusion.

**Tets plan:** New unit test

### CODE REVIEW

**Review by ajkr — APPROVED** [review]
Yeah I think the formula makes sense. LGTM

File: db/compaction/compaction_picker_universal.cc [hunk 1]
```diff
@@ -129,6 +129,94 @@ class UniversalCompactionBuilder {
...
+    const std::size_t num_l0_to_exclude_for_max_merge_width =
+        std::min(max_merge_width -
+                     std::min(max_merge_width, num_extra_l0_before_write_stop),
+                 max_num_l0_to_exclude);
+    const std::size_t num_l0_to_exclude_for_min_merge_width =
+        std::min(min_merge_width -
+                     std::min(min_merge_width, num_extra_l0_before_write_stop),
+                 max_num_l0_to_exclude);
+
+    std::size_t num_l0_to_exclude = 0;
```
[the remark this answers is not in the document]
Comment (hx235) [reply db/compaction/compaction_picker_universal.cc:188]: Pending discussion offline.

File: db/compaction/compaction_picker_universal.cc [hunk 2]
```diff
@@ -129,6 +129,94 @@ class UniversalCompactionBuilder {
...
+        mutable_cf_options_.compaction_options_universal
+            .max_size_amplification_percent);
+    const size_t base_sr_size = sorted_runs_[end_index].size;
+
+    // Leave at least 1 L0 file and 2 input sorted runs after exclusion
+    const std::size_t max_num_l0_to_exclude =
+        std::min(num_l0_input_pre_exclusion - 1, end_index - start_index - 1);
+    const std::size_t num_extra_l0_before_write_stop =
+        level0_stop_writes_trigger -
+        std::min(level0_stop_writes_trigger, num_l0_input_pre_exclusion);
```
[the remark this answers is not in the document]
Comment (hx235) [reply db/compaction/compaction_picker_universal.cc:178]: Thanks - I believe we can do 
```
// In universal compaction, sorted runs from non L0 levels are counted
    // toward `level0_stop_writes_trigger`. Therefore we need to subtract the
    // total number of sorted runs picked originally for this compaction from
    // `level0_stop_writes_trigger` to calculate
    // `num_extra_l0_before_write_stop`
    const std::size_t num_extra_l0_before_write_stop =
        level0_stop_writes_trigger -
        std::min(level0_stop_writes_trigger, end_index - start_index + 1);
```

Will fix this and the UT

File: unreleased_history/behavior_changes/exclude_some_l0_size_amp.md [hunk 3]
```diff
@@ -0,0 +1 @@
+Universal size amp compaction will conditionally exclude some L0 files (newest first) when selecting input with a small negative impact to size amp. This is to prevent a large number of L0 files from being locked by a size amp compaction, potentially leading to write stop with a few more flushes.
```
[the remark this answers is not in the document]
Comment (hx235) [reply unreleased_history/behavior_changes/exclude_some_l0_size_amp.md]: Fixed

File: db/compaction/compaction_picker_universal.cc [hunk 4]
```diff
@@ -800,6 +888,7 @@ Compaction* UniversalCompactionBuilder::PickCompactionToReduceSizeAmp() {
   const size_t base_sr_size = sorted_runs_[end_index].size;
   size_t start_index = end_index;
   size_t candidate_size = 0;
+  size_t l0_candidate_count = 0;
```
[the remark this answers is not in the document]
Comment (hx235) [reply db/compaction/compaction_picker_universal.cc]: Fixed

File: db/compaction/compaction_picker_universal.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (hx235) [reply db/compaction/compaction_picker_universal.cc]: Fixed the ones related to file system size

**Pull request closed without merging**